dolphinscheduler-task-plugin/dolphinscheduler-task-emr/src/main/java/org/apache/dolphinscheduler/plugin/task/emr/EmrTask.java

```diff
@@ -0,0 +1,249 @@
+        if (emrParameters == null || !emrParameters.checkParameters()) {
+            throw new EmrTaskException("emr task params is not valid");
+        }
+        emrClient = createEmrClient();
+    }
+
+    @Override
+    public void handle() throws InterruptedException {
+        ClusterStatus clusterStatus = null;
+        try {
+            RunJobFlowRequest runJobFlowRequest = createRunJobFlowRequest();
+
+            // submit runJobFlowRequest to aws
+            RunJobFlowResult result = emrClient.runJobFlow(runJobFlowRequest);
+
+            clusterId = result.getJobFlowId();
+
+            clusterStatus = getClusterStatus();
+
+            while (waitingStateSet.contains(clusterStatus.getState())) {
+                TimeUnit.SECONDS.sleep(10);
+                clusterStatus = getClusterStatus();
+            }
+
+        } catch (EmrTaskException | SdkBaseException e) {
+            logger.error("emr task submit failed with error", e);
+            setExitStatusCode(TaskConstants.EXIT_CODE_FAILURE);
+        } finally {
+            if (clusterStatus != null) {
+                calculateExitStatusCode(clusterStatus);
+                logger.info("emr task finished with cluster status : {}", clusterStatus);
+            }
+        }
+    }
+
+    /**
+     * parse json string to RunJobFlowRequest
+     *
+     * @return RunJobFlowRequest
+     */
+    private RunJobFlowRequest createRunJobFlowRequest() {
+
+        final RunJobFlowRequest runJobFlowRequest;
+        try {
+            runJobFlowRequest = objectMapper.readValue(emrParameters.getJobFlowDefineJson(), RunJobFlowRequest.class);
+        } catch (JsonProcessingException e) {
+            throw new EmrTaskException("can not parse RunJobFlowRequest from json", e);
+        }
+
+        return runJobFlowRequest;
+    }
+
+    /**
+     * calculate task exitStatusCode
+     *
+     * @param clusterStatus aws emr cluster status
+     */
+    private void calculateExitStatusCode(ClusterStatus clusterStatus) {
+        String state = clusterStatus.getState();
+        ClusterStateChangeReason stateChangeReason = clusterStatus.getStateChangeReason();
+        ClusterState clusterState = ClusterState.valueOf(state);
+        int exitStatusCode;
+        switch (clusterState) {
+            case TERMINATED:
+            case TERMINATING:
+                String code = stateChangeReason.getCode();
+                if (code != null && code.equalsIgnoreCase(ClusterStateChangeReasonCode.ALL_STEPS_COMPLETED.toString())) {
+                    exitStatusCode = TaskConstants.EXIT_CODE_SUCCESS;
+                } else {
+                    exitStatusCode = TaskConstants.EXIT_CODE_KILL;
+                }
+                break;
+            case TERMINATED_WITH_ERRORS:
+                exitStatusCode = TaskConstants.EXIT_CODE_FAILURE;
+                break;
+            default:
+                exitStatusCode = TaskConstants.EXIT_CODE_SUCCESS;
```

Comment:
Is it OK of default success? If other state add in the future, it will be wrong. What do you think?

Reply:
accept it